### ISSUE TITLE
Reuse already loaded assemblies if they are identical

### ISSUE THREAD

**Issue by tmat** [dotnet/roslyn#6474]
Scenario:
Solution with 3 projects that don't build into a common directory but rather each builds into their own:

```
Shared\bin\debug\Shared.dll
Lib1\bin\debug\Lib1.dll 
               Shared.dll (copy)
Lib2\bin\debug\Lib2.dll
               Shared.dll (copy)
```

Now in a REPL:

```C#
> #r "Lib1\bin\debug\Lib1.dll" 
> Lib1 l1 = new Lib1();  // loads Lib1 and Shared from Lib1\bin\debug dir
> #r "Lib2\bin\debug\Lib2.dll" 
> Lib2 l2 = new Lib2();   // loads Lib1 and Shared from Lib2\bin\debug dir
> l2.libBase
Field not found: 'Lib2.libBase'.
  + System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[TResult].Start(TStateMachine& stateMachine)
```

The problem is that we load Shared twice at runtime because it's in a different directory. The compiler correctly dedups it since both copies have the same identity. The loader should check whether the assemblies are the same. If they are then we can reuse the existing one. Otherwise we should report an error.

**Comment by tmat** [comment]
FYI @genlu @ManishJayaswal

**Comment by jasonwilliams200OK** [comment]
Seems like a typo at `> #r "Lib1\bin\debug\Lib2.dll"`. Shouldn't it be `Lib2\bin`?

**Comment by tmat** [comment]
@jasonwilliams200OK Yes, Thanks

**Comment by tmat** [comment]
Fixed by https://github.com/dotnet/roslyn/commit/778c54d82b8de7b1936ca269d346d774b2bd33b7
